fuse_loss/include/fuse_loss/loss_function.h

```diff
@@ -0,0 +1,191 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_LOSS_LOSS_FUNCTION_H
+#define FUSE_LOSS_LOSS_FUNCTION_H
+
+#include <ceres/loss_function.h>
+
+// This provides additional loss functions that are not available in:
+// https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/loss_function.h
+//
+// Most of them are available in other libraries like:
+// * g2o: https://github.com/RainerKuemmerle/g2o/blob/master/g2o/core/robust_kernel_impl.h
+// * GTSAM: https://github.com/borglab/gtsam/blob/develop/gtsam/linear/LossFunctions.h
+//          (formerly in https://github.com/borglab/gtsam/blob/develop/gtsam/linear/NoiseModel.h)
+//
+// Remember that the loss functions in Ceres solver implement a slight variation of the \rho(r) function, defined for
+// multiple M-estimators in table #1 from http://www.audentia-gestion.fr/research.microsoft/ZhangIVC-97-01.pdf (p. 24).
+//
+// In Ceres the loss functions evaluate:
+//
+//   rho(s) = 2 * \rho(sqrt(s))
+//
+// where s = r^2, being r the residual.
+//
+// This is because the output of rho(s) is mutiplied by 0.5 when the residual block cost is computed, either with or
+// without loss:
+// * cost w/o loss: https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/residual_block.cc#L159
+// * cost w/  loss: https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/residual_block.cc#L165
+//
+// Note that according to this, it looks like the following loss functions are incorrectly implemented in Ceres:
+// * Tukey: It must be multiplied by 2, so instead of dividing by 6 it should divide by 3. See:
+//
+//            https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/loss_function.h#L281-L282
+//
+//          There is an easy workaround for this: combine TukeyLoss with ScaledLoss, using a scaled factor of 2.
+//
+// * Huber: It is actually a Pseudo-Huber loss function, which is smoother than the original Huber one in table #1 from
+//          http://www.audentia-gestion.fr/research.microsoft/ZhangIVC-97-01.pdf (p. 24).
+//
+//          It also does not look to match the Pseudo-Huber variation suggested in:
+//
+//            http://en.wikipedia.org/wiki/Huber_Loss_Function
+//
+//          which is referenced in:
+//
+//            https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/loss_function.h#L173
+namespace ceres
+{
+
+// Dynamic Covariance Scaling (DCS), equivalent to the Geman-McClure with the tuning constant 'a', but scaled by 2.0.
+//
+// The term is computed as:
+//
+//   rho(s) = s * 2 * a / (a + s)    for s <  a
+//   rho(s) = s                      for s >= a
+//
+// See http://www2.informatik.uni-freiburg.de/~spinello/agarwalICRA13.pdf (p. 3), where 's' is multiplied by the factor:
+//
+//   min(1, 2 * a / (a + s))
+//
+// The g2o implementation seems to use the squared of that factor:
+// https://github.com/RainerKuemmerle/g2o/blob/master/g2o/core/robust_kernel_impl.cpp#L167
```

Comment:
I believe I'm doing the right thing, and either g2o implementation is:
1. Wrong: it doesn't match the explanation on the paper BUT the resulting loss function is good, but different
2. Wrong: it doesn't match the explanation on the paper AND the resulting loss function is bad
3. Right: it simply requires the scale factor squared because of the way the result is used in the rest of the system, e.g. used to multiplied the information matrix, instead of the residuals.

Not sure what's the right answer here or how to validate which one is the correct implementation. I tried plotting the loss function using the scale factor as is and squared, and it looks to me that it shouldn't be squared. I don't have any reference plot to compare with, but with the scale factor as is, it looks closer to **Geman-McClure**.

Reply:
In the paper, the scale factor s_ij is multiplied with the error term, before being squared.
For example, see Eq (6):
... || s_ij * (f(x_i, x_j) - z_ij) ||^2 ...

That would imply to me that the scale factor should be squared before multiplying with `s`, the Ceres squared residual  term.

Looking further on in Eq (6), you see they reduce the equation to: s_kl^2 * X^2
Again, this points towards the scale factor squared times the residual squared being the intent of the paper.

And justbefore the start of Section 4-A on pg4:
> the residuals can be scaled with s_ij or the information matrix be scaled with s_ij^2 depending on implementation

And since this paper was written by Wolfram Burgard, g2o was also written by Wolfram Burgard, and the paper says this scale factor was tested with g2o, I would suspect that g2o contains the reference implementation.

Reply:
Thanks for taking the time to look at the paper. I think you're right, `s_kl` in eq. (6) represents the switchable constraint for the edge `kl`. Since that point in the paper, they simply call it `s`, and compute a first intermediate value in eq. (8) and substitute it in `h()`, which is defined as part of eq. (6). I haven't verified the derivative of `\hat{h}`, but I'm going to assume it's correct for now.

So I agree, from the paper, that we should be scaling by the squared, since in Ceres we're multiplying the squared residual.

However, when I plot this approach (let's call it g2o) vs. the one I implemented initially, I get different functions. That's expected, but I'm not sure if it's correct that the weight function is negative for the outlier region. If you think that's ok, then I'd make the change. It's just that the weight function jumps to 0, then goes up (in absolute value) and it finally converges to zero, which is the only part that makes real sense to me.

The following table compares the two implementations. My main concern can be observed in the DCS g2o absolute weight plot, which is what I tried to described in the previous paragraph. What do you think?

| Function | DCS | DCS g2o |
| ------------ | ------ | ------------ |
| `\rho` | ![rho_dcs](https://user-images.githubusercontent.com/382167/75160233-f4bc4700-5719-11ea-95eb-2b029c7fd9b7.png) | ![rho_dcs_g2o](https://user-images.githubusercontent.com/382167/75160247-f980fb00-5719-11ea-972e-265f2c0639c3.png) |
| Influence (`\phi`) | ![influence_dcs](https://user-images.githubusercontent.com/382167/75160265-fede4580-5719-11ea-897c-7fac45164daa.png) | ![influence_dcs_g2o](https://user-images.githubusercontent.com/382167/75160278-03a2f980-571a-11ea-89ea-2d3afcc62b10.png) |
| Weight | ![weight_dcs](https://user-images.githubusercontent.com/382167/75160289-0867ad80-571a-11ea-88a7-2794ec7a8402.png) | ![weight_dcs_g2o](https://user-images.githubusercontent.com/382167/75160299-0d2c6180-571a-11ea-9495-01e7968563df.png) |
| Absolute Weight | ![absolute_weight_dcs](https://user-images.githubusercontent.com/382167/75160315-14536f80-571a-11ea-849d-8f01305f02cd.png) | ![absolute_weight_dcs_g2o](https://user-images.githubusercontent.com/382167/75160326-17e6f680-571a-11ea-9950-17d46098b7a1.png) |

To put the DCS g2o approach into context, the following table has it together with the other loss functions:

| Function | Plot |
|----| ----|
|`\rho`| ![rho](https://user-images.githubusercontent.com/382167/75160479-55e41a80-571a-11ea-84a3-d320a9f5ec73.png) |
|Influence (`\phi`) | ![influence](https://user-images.githubusercontent.com/382167/75160469-51b7fd00-571a-11ea-862f-5416bf50228c.png)|
| Weight | ![weight](https://user-images.githubusercontent.com/382167/75160487-5b416500-571a-11ea-8ce3-270ad3c0f104.png)|
|Absolute Weight | ![absolute_weight](https://user-images.githubusercontent.com/382167/75160459-4d8bdf80-571a-11ea-87a9-3235999e17b5.png) |

Reply:
@svwilliams I've addressed all your comments, including this one. At this moment, the PR uses the squared of the scale factor suggested in the paper, as **g2o** does.

However, I left this conversation unresolved, so you can see how the plots look like (the weight and absolute weight in particular) and comment. If you think this is the right thing, you can merge this PR, since it already has that approach implemented. Otherwise, I can revert the commit back.

Another review pass would be required though, to confirm all comments have been updated and they're consistent. I did a passed myself and I believe they're, but I could have missed some.

Reply:
I just gave the DCS **g2o** implementation a try and I got a crashed because the derivative must be positive:

``` bash
WARNING: Logging before InitGoogleLogging() is written to STDERR
F0224 18:28:51.008424  9942 corrector.cc:93] Check failed: rho[1] > 0.0 (-7.18086e-06 vs. 0) 
```

This error message is likely part of an `assert` because right after that the node crashed.

So far I was simply running the unit test.

`rho[1]` is negative by a small margin though, so I'm not completely sure the implementation is wrong, or not acceptable according to some Ceres invariants that loss functions should satisfy, or maybe it's simply a numeric error. :thinking:

Reply:
I've done all the math again to confirm the derivatives I've implemented are correct.

In **g2o** the derivatives are wrong: https://github.com/RainerKuemmerle/g2o/blob/fcba4eaca6f20d9a5792404cc8ef303aeb8ba5d2/g2o/core/robust_kernel_impl.cpp#L162-L169

In order for `rho[1]` to be `scale * scale`, `rho[0]` should be `2 * scale * s`. And the second derivative isn't `0`, but `rho[2] = - 2 * rho[1] / (phi + e2)`, where `phi + e2` would be `a + s` in my DCS implementation.

So **g2o** is basically using the squared of the scale factor in the paper for `rho[0]`, but it computes the first derivative as if `rho[0]` had been computed with the scale factor (not squared, and multiplied by `2`), and the second derivative is just set to `0`. :unamused:

Reply:
For the current implementation in this PR, using the squared scale factor as in **g2o**, but the correct derivatives, I get:

- `rho[2] > 0` for `s > a/2`
- `rho[1] < 0` for `s > a`

When this happens, the corrector fails.

Since the outlier region is indeed defined for `s > a`, that means the corrector will always fail for any outlier, so somehow the DCS loss function with the squared scale factor is invalid. Or I'm missing something else.

Reply:
I read the g2o implementation as pretending the scale term was not a function of s.
`rho[0] = scale * scale * s;`
If you pretend `scale` is a constant, then d/ds( scale * scale * s) = scale * scale
`rho[1] = scale * scale;`
And then the derivative of a constant is zero.
`rho[2] = 0;`

Not that I think this is correct, because `scale` is a function of s.

I don't really have any suggestions right now. I'll poke at the derivatives a bit later, but I'm leaning towards "the DCS formulation is not a valid loss function" being the answer.

Reply:
> I'm leaning towards "the DCS formulation is not a valid loss function" being the answer

 Me too. It's a shame in the paper they didn't plot the `rho`-function or the weight (the `rho` first derivative), so at least I get a better idea of what they actually used. I'd assume it was the **g2o** implementation as you mentioned before.

Yep, it looks like they didn't apply the chain rule for the `scale` derivative, so they treat it as a constant in **g2o**.

At this point, since this PR also contributes other loss functions, I wonder what would be an acceptable path forward to get this PR in:

- [ ] Leave the DCS with the squared scale factor, but hack a zero 2nd derivative `rho[2] = 0;`, and `DISABLED_` the unit test for the DCS loss function.
- [ ] Revert the DCS to the implementation that uses the non-squared scale factor, and add a comment explaining why we don't use the squared scale factor, although according to the paper and **g2o** implementation we should be using the squared scale factor.
- [ ] Remove the DCS loss function and leave a comment in the plugins failing explaining why the DCS is not provided. This would be a shame though, because it's available in **g2o** and also **GTSAM**: https://github.com/borglab/gtsam/blob/09b0f03542bfbec5cca62645a60c5d1d4f8fc48c/gtsam/linear/LossFunctions.cpp#L348-L357

That being said, I also wonder if the DCS loss function is not a valid loss function / M-estimator, or it's a type not supported by Ceres corrector. I wish I had a reference book that enunciates the properties an M-estimator should satisfy. I guess I could look at the techniques mentioned in the corrector code, since they should clarify why `rho[1]` must be positive when `rho[2]` is positive. But it's indeed better to have loss functions where `rho[2]` is negative, because that runs a faster path in the corrector, from the little I read in the Ceres implementation file.

Reply:
I'm curious what happens if you use the g2o derivatives for both the first and second derivative. Obviously it won't pass the finite difference unit tests, but I'm curious if Ceres will freak out. Having the derivatives point in an inaccurate direction may not result in anything good.

If some variation of that doesn't work, I say comment it out.

Either way, you should add a big warning to the DCS header.

Reply:
I think it'd be worth taking **GTSAM** implementation as a reference as well: https://github.com/borglab/gtsam/blob/develop/gtsam/linear/LossFunctions.cpp

I think the `residual` method should be equivalent to `rho[0]` and the `weight` method should be equivalent to `rho[1]`.

IIRC I contributed DCS to **GTSAM** a while ago, but it looks like the noise models have been refactored and changed significantly since them. Indeed, I don't remember there was a separate `weight` and `residual` method, and the `residual` method implementation doesn't look familiar to me, so it was contributed by someone else.

Indeed, this was my contribution: https://github.com/borglab/gtsam/commit/b9c63ef5df83767b055b2fb4165100dffc139dd8

By then, I only had to implement the `weight` method, which I believe I did correctly. :thinking: 

The noise models got separated into a loss function file about 3 months ago, making it harder to blame how contributed the `residuals` method implementation: https://github.com/borglab/gtsam/commit/73542d980ab6cef78a2866b404ac01d05d6178a5

The `residual` method was added in https://github.com/borglab/gtsam/commit/431c37090a1e133ee400e5316cfd206cf792d82f by https://github.com/varunagrawal

Reply:
**GTSAM** in https://github.com/borglab/gtsam/blob/develop/gtsam/linear/LossFunctions.cpp implements the `\rho(x)` (`residual` method) and `w(x)` (`weight` method) equations from table (1) in http://www.audentia-gestion.fr/research.microsoft/ZhangIVC-97-01.pdf (p. 24).

We have that:

- `\rho(x)` is the `residual` method
- `\phi(x) = d \rho(x) / dx`
- `w(x) = \phi(x) / x` is the `weight` method

In **Ceres** we should have the following equivalences in `Evaluate(s)`, where `s = x^2` (`s` being the squared residual, and `x` the residual):

- `rho(s)[0] = 2 * residual(sqrt(s))`, so `residual(x) = 0.5 * rho(x^2)[0]`
- `rho(s)[1] = weight(sqrt(s))` because the 1st derivative of `residual(x)` is `0.5 * rho(x^2)[1] * 2 * x = rho(x^2)[1] * x = \phi(x)`. So `rho(x^2)[1] = \phi(x) / x = w(x)`

A good example is the **Cauchy** loss function:

- Ceres: https://github.com/ceres-solver/ceres-solver/blob/13c7a22cee789fa03a72520b1cb861cd8e9a5a54/internal/ceres/loss_function.cc#L77-L78
- GTSAM: https://github.com/borglab/gtsam/blob/09b0f03542bfbec5cca62645a60c5d1d4f8fc48c/gtsam/linear/LossFunctions.cpp#L211-L218

If we take the DCS implementation in GTSAM: https://github.com/borglab/gtsam/blob/09b0f03542bfbec5cca62645a60c5d1d4f8fc48c/gtsam/linear/LossFunctions.cpp#L348-L367, we have:

- `residual(x) = (a^2 * x^2 + a * x^4) / (a + x^2)^2` that can be further simplified into `a * x^2 * (a + x^2)  / (a + x^2)^2 = a * x^2 / (a + x^2)`
- `weight(x) = (2 * a / (a + x^2))^2`

I've tried to confirm `weight(x) = d residual(x) / dx * x`, but it's not exactly the same, but almost. I get `weight(x) = 2 * (a / (a + x^2))^2`.

In order to get `weight(x) = (2 * a / (a + x^2))^2`, we simply need `residual(x) = 2 * a * x^2 / (a + x^2)`.

Therefore, in **Ceres** we'd have:

- `rho(s)[0] = 2 * residual(sqrt(s)) = 4 * a * s / (a + s)`
- `rho(s)[1] = weight(sqrt(s)) = (2 * a / (a + s))^2`

In other words, the `scale` factor for `rho(s)[0]` would be `4 * a / (a + s)`.

This would be very similar to the **Geman-McClure** loss function: https://github.com/borglab/gtsam/blob/09b0f03542bfbec5cca62645a60c5d1d4f8fc48c/gtsam/linear/LossFunctions.cpp#L314-L325. This is what **GTSAM** has for it:

- `residual(x) = 0.5 * (a^2 * e^2) / (a^2 + e^2)`
- `weight(x) = a^4/(a^2 + e^2)^2`

That in **Ceres** would be:

- `rho(s)[0] = (a^2 * s) / (a^2 + s)`
- `rho(s)[1] = a^4/(a^2 + s)^2`

This is certainly different that **DCS**, but my implementation of **Geman-McClure** in this PR doesn't match that.

It looks like I could do the following:
1. Update the **DCS** implementation to the one described before (similar to **GTSAM**, expect for an additional factor of `2`)
2. Change the **Geman-McClure** implementation to match the one describe before. Note that **GTSAM** implementation actually matches the equations from table (1) in http://www.audentia-gestion.fr/research.microsoft/ZhangIVC-97-01.pdf (p. 24) if we replace `1` with `a^2` in the `w(x)` function. If we replace `x` with `x/a` in `\rho(x)` we don't get the `a^2` though.

Even if **DCS** and **Geman-McClure** are similar, at least with **DCS** there's an inlier and outlier region, so it still makes sense to have both. If I do what I've mentioned before, I should be able to get something that works, since it's very similar to the original implementation I sent to this PR. Does this make sense and work for you, @svwilliams ?

Reply:
That works for me.

Reply:
I found http://www2.informatik.uni-freiburg.de/~agarwal/resources/agarwal-thesis.pdf and I think now everything makes sense.

I've updated the DCS `rho[0]` equation accordingly and the comments in header, implementation and plugin files. You can see and review the details there.

I also rebased the commits.

This is how the loss functions look like now, including DCS with the last implementation (sorry it's in yellow):

|Function|Plot|
|-|-|
|`\rho(r)`|![rho](https://user-images.githubusercontent.com/382167/75334791-be510a00-5888-11ea-903d-e46a50ea7be5.png)|
|Influence: `\rho'(r) * r`| ![influence](https://user-images.githubusercontent.com/382167/75334837-cf9a1680-5888-11ea-833b-5135d64d07d2.png)|
|Weight: `\rho'(r)`|![weight](https://user-images.githubusercontent.com/382167/75334878-e17bb980-5888-11ea-8eb1-e13e5675b69a.png)|
|`\rho''(r)`|![second_derivative](https://user-images.githubusercontent.com/382167/75334908-ec364e80-5888-11ea-9a5d-a0b77f269e6f.png)|